src/mono/mono/utils/mono-threads-wasm.c

```diff
@@ -311,6 +311,9 @@ mono_threads_platform_in_critical_region (THREAD_INFO_TYPE *info)
 void
 mono_memory_barrier_process_wide (void)
 {
+#ifndef DISABLE_THREADS
+	mono_memory_barrier ();
```

Comment:
Maybe this is not enough for GC. Maybe we need to set barrier on all threads ? How?